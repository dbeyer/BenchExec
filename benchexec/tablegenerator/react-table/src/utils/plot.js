// This file is part of BenchExec, a framework for reliable benchmarking:
// https://github.com/sosy-lab/benchexec
//
// SPDX-FileCopyrightText: 2019-2020 Dirk Beyer <https://www.sosy-lab.org>
//
// SPDX-License-Identifier: Apache-2.0

import React from "react";

/**
 * Renders a setting (= a dropdown menu with its label) for one of the plots.
 *
 * @param {String} name name of the dropdown that will be used for the label next to it
 * @param {String} value default value that will be selected in the dropdown
 * @param {function} changeHandler handler function that will be called when an option was selected
 * @param {Object} options object containing the names of all options for the dropdown
 * @param {String} tooltip [OPTIONAL] tooltip for the whole setting
 * @param {boolean} isDisabled [OPTIONAL] whether or not the dropdown is disabled
 **/
const renderSetting = (
  name,
  value,
  changeHandler,
  options,
  tooltip,
  isDisabled,
) => {
  return (
    <div className={`setting${isDisabled ? " disabled" : ""}`} title={tooltip}>
      <span className={`setting-label${tooltip ? " with-tooltip" : ""}`}>
        {name}:
      </span>
      <select
        className="setting-select"
        name={"setting-" + name}
        value={isDisabled ? "disabled" : value}
        onChange={changeHandler}
        disabled={isDisabled}
      >
        {Object.values(options).map((option) => (
          <option value={option} key={option} name={option + " " + name}>
            {option}
          </option>
        ))}
        {isDisabled ? (
          <option value="disabled" name="disabled">
            â¸º
          </option>
        ) : (
          ""
        )}
      </select>
    </div>
  );
};

/**
 * Renders a button to reset all modifications applied to plot.
 *
 * @param {function} resetHandler handler function triggered by button click to reset plot modifications
 **/
const renderResetButton = (resetHandler) => {
  return (
    <button className="setting-button" onClick={() => resetHandler()}>
      Reset plot
    </button>
  );
};

/**
 * Renders a setting (= a dropdown menu with its label) for one of the plots.
 *
 * @param {String} name name of the dropdown that will be used for the label next to it
 * @param {String} value default value that will be selected in the dropdown
 * @param {function} changeHandler handler function that will be called when an option was selected
 * @param {Object} options object containing the name of the optgroup as key and an array of objects representing the selections
   with their display name and their value property
 * @param {String} tooltip [OPTIONAL] tooltip for the whole setting
 **/
const renderOptgroupsSetting = (
  name,
  value,
  changeHandler,
  options,
  tooltip,
) => {
  return (
    <div className="setting" title={tooltip}>
      <span className={`setting-label${tooltip ? " with-tooltip" : ""}`}>
        {name}:
      </span>
      <select
        id={"setting-" + name}
        className="setting-select"
        name={"setting-" + name}
        value={value}
        onChange={changeHandler}
      >
        {Object.entries(options).map(([optgroup, selections]) => (
          <optgroup label={optgroup} key={optgroup}>
            {selections.map((optionObj) => (
              <option
                value={optionObj.value}
                key={optionObj.value}
                name={optionObj.name + " " + name}
              >
                {optionObj.name}
              </option>
            ))}
          </optgroup>
        ))}
      </select>
    </div>
  );
};

/**
 * Returns an object containing arrays of data points (of the format [x-value, y-value]) that represent the upper and lower
 * borders of the 95% confidence interval of the linear regression line. Calculation is done via the formula provided by
 * https://www2.stat.duke.edu/courses/Spring13/sta101.001/slides/unit6lec3H.pdf
 * @param {Array} actualData array of actual data points that were used for the linear regression
 * @param {Array} predictedData array of calculated data points that represent the linear regression line
 * @param {function} regressionFunction function that calculates the predicted values
 * @param {int} minX leftmost x value for the borders
 * @param {int} maxX rightmost x value for the borders
 */
function getConfidenceIntervalBorders(
  actualData,
  predictedData,
  regressionFunction,
  minX,
  maxX,
) {
  const getXValue = (data) => data[0];
  const getYValue = (data) => data[1];
  const sum = (x, y) => x + y;
  minX = Math.floor(minX);
  maxX = Math.ceil(maxX);
  // Value of the t-statistic for a 95% confidence interval
  const tValue = 1.96;
  const stdErr = Math.sqrt(
    actualData
      .map((data, index) => [getYValue(data), getYValue(predictedData[index])])
      .map((yValues) => Math.pow(yValues[1] - yValues[0], 2))
      .reduce(sum) / actualData.length,
  );
  const meanOfX =
    actualData.map((data) => getXValue(data)).reduce(sum) / actualData.length;
  const stdOfX = Math.sqrt(
    actualData
      .map((data) => Math.pow(getXValue(data) - meanOfX, 2))
      .reduce(sum) / actualData.length,
  );

  const dataPointsOfRegression = getDataPointsOfRegression(
    minX,
    maxX,
    regressionFunction,
  );

  const diffMargins =
    stdErr === 0 || stdOfX === 0
      ? dataPointsOfRegression.map((data) => 0)
      : dataPointsOfRegression.map((data) =>
          Number.parseFloat(
            (
              tValue *
              stdErr *
              Math.sqrt(
                1 / actualData.length +
                  Math.pow(getXValue(data) - meanOfX, 2) /
                    ((actualData.length - 1) * Math.pow(stdOfX, 2)),
              )
            ).toPrecision(4),
          ),
        );

  return {
    upperBorderData: dataPointsOfRegression.map((data, index) => [
      getXValue(data),
      getYValue(data) + diffMargins[index],
    ]),
    lowerBorderData: dataPointsOfRegression.map((data, index) => [
      getXValue(data),
      getYValue(data) - diffMargins[index],
    ]),
  };
}
/* Computes the data points that will be plotted for a regression. The maximum
   number of data points will not exceed 10.000. If there were more, the lines
   will be plotted in intervals. */
function getDataPointsOfRegression(minX, maxX, regF) {
  const thresholds = [100000000, 10000000, 1000000, 100000, 10000];
  const threshold = thresholds.find((threshold) => maxX > threshold);
  const numberPointsDivider = threshold ? threshold / 1000 : 1;
  const dataPointsOfRegression = Array(Math.ceil(maxX / numberPointsDivider))
    .fill()
    .map((x, index) => index * numberPointsDivider);

  return dataPointsOfRegression
    .filter((int) => int >= minX)
    .map((number) => {
      const pred = regF(number);
      return [
        Number.parseFloat(pred[0].toPrecision(4)),
        Number.parseFloat(pred[1].toPrecision(4)),
      ];
    });
}

export {
  renderSetting,
  renderOptgroupsSetting,
  renderResetButton,
  getConfidenceIntervalBorders,
  getDataPointsOfRegression,
};
