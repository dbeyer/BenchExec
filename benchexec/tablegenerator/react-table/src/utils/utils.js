/* SPDX-License-Identifier: Apache-2.0
 *
 * BenchExec is a framework for reliable benchmarking.
 * This file is part of BenchExec.
 * Copyright (C) Dirk Beyer. All rights reserved.
 */
import React from "react";

const prepareTableData = ({head, tools, rows, stats, props}) => {
  return {
    tableHeader: head,
    tools: tools.map(tool => ({
      ...tool,
      isVisible: true,
      columns: tool.columns.map(column => ({...column, isVisible: true})),
    })),
    columns: tools.map(tool => tool.columns.map(column => column.title)),
    table: rows,
    stats,
    properties: props,
  };
};

const isNumericColumn = column => column.type === "count" || column.type === "measure";

const applyNumericFilter = (filter, row, cell) => {
  const raw = getRawOrDefault(row[filter.id]);
  if (raw === undefined) {
    // empty cells never match
    return false;
  }
  const filterParams = filter.value.split(":");

  if (filterParams.length === 2) {
    const [start, end] = filterParams;

    const numRaw = Number(raw);
    const numStart = Number(start);
    const numEnd = end ? Number(end) : Infinity;

    return numRaw >= numStart && numRaw <= numEnd;
  }

  if (filterParams.length === 1) {
    return raw.startsWith(filterParams[0]);
  }
  return false;
};

const applyTextFilter = (filter, row, cell) => {
  const raw = getRawOrDefault(row[filter.id]);
  if (raw === undefined) {
    // empty cells never match
    return false;
  }
  return raw.includes(filter.value);
};

const isNil = data => data === undefined || data === null;

const getRawOrDefault = (value, def) => (isNil(value) || isNil(value.raw) ? def : value.raw);

const numericSortMethod = (a, b) => {
  const aValue = getRawOrDefault(a, +Infinity);
  const bValue = getRawOrDefault(b, +Infinity);
  return aValue - bValue;
};

const textSortMethod = (a, b) => {
  const aValue = getRawOrDefault(a, "").toLowerCase();
  const bValue = getRawOrDefault(b, "").toLowerCase();
  if (aValue === "") {
    return 1;
  }
  if (bValue === "") {
    return -1;
  }
  if (aValue > bValue) {
    return 1;
  }
  if (aValue < bValue) {
    return -1;
  }
  return 0;
};

const isOkStatus = status => {
  return status === 0 || status === 200;
};

// Best-effort attempt for calculating a meaningful column width
const determineColumnWidth = (column, minWidth, maxWidth) => {
  let width = column.max_width; // number of chars in column
  if (minWidth) {
    width = Math.max(width, minWidth);
  }
  if (maxWidth) {
    width = Math.min(width, maxWidth);
  }
  if (!width) {
    width = 10;
  }

  return width * 8 + 20;
};

const formatColumnTitle = column =>
  column.unit ? (
    <>
      {column.display_title}
      <br />
      {`(${column.unit})`}
    </>
  ) : (
    column.display_title
  );

const getRunSetName = ({tool, date, niceName}) => {
  return `${tool} ${date} ${niceName}`;
};

// Extended color list copied from
// https://github.com/uber/react-vis/blob/712ea622cf12f17bcc38bd6143fe6d22d530cbce/src/theme.js#L29-L51
// as advised in https://github.com/uber/react-vis/issues/872#issuecomment-404915958
const EXTENDED_DISCRETE_COLOR_RANGE = [
  "#19CDD7",
  "#DDB27C",
  "#88572C",
  "#FF991F",
  "#F15C17",
  "#223F9A",
  "#DA70BF",
  "#125C77",
  "#4DC19C",
  "#776E57",
  "#12939A",
  "#17B8BE",
  "#F6D18A",
  "#B7885E",
  "#FFCB99",
  "#F89570",
  "#829AE3",
  "#E79FD5",
  "#1E96BE",
  "#89DAC1",
  "#B3AD9E",
];

export {
  prepareTableData,
  getRawOrDefault,
  isNumericColumn,
  applyNumericFilter,
  applyTextFilter,
  numericSortMethod,
  textSortMethod,
  determineColumnWidth,
  formatColumnTitle,
  getRunSetName,
  isOkStatus,
  isNil,
  EXTENDED_DISCRETE_COLOR_RANGE,
};
