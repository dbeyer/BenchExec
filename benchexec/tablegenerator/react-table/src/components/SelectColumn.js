/* SPDX-License-Identifier: Apache-2.0
 *
 * BenchExec is a framework for reliable benchmarking.
 * This file is part of BenchExec.
 * Copyright (C) Dirk Beyer. All rights reserved.
 */
import React from "react";
import ReactModal from "react-modal";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faTimes } from "@fortawesome/free-solid-svg-icons";
import { getRunSetName } from "../utils/utils";

export default class SelectColumn extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      deselect: true,
      list: [...this.props.tools]
    };
    this.selectable = [];
  }

  // -------------------------Rendering-------------------------
  renderRunSets = () => {
    return this.state.list.map((tool, i) => {
      let isVisible =
        this.state.list[i].columns.findIndex(
          value => value.isVisible === true
        ) > -1;
      let toolName = getRunSetName(tool);
      return (
        <tr id={toolName} key={"tr" + toolName}>
          <td
            id={toolName}
            key={"key" + toolName}
            className={isVisible ? "checked" : ""}
          >
            <label>
              {toolName}
              <input
                name={toolName}
                type="checkbox"
                checked={isVisible}
                onChange={e => this.deselectTool(i, e)}
              ></input>
            </label>
          </td>
          {this.renderColumns(i)}
        </tr>
      );
    });
  };

  renderColumns = index => {
    const self = this;
    const idxColumn = [];
    const length = this.state.list[index].columns.length;
    this.selectable.forEach(function(headerRow, idxHeader) {
      self.state.list[index].columns.forEach(function(column, idx) {
        if (headerRow.display_title === column.display_title) {
          idxColumn.push({
            display_title: column.display_title,
            isVisible: column.isVisible,
            key: idxHeader
          });
        } else if (
          idx === length - 1 &&
          idxColumn
            .map(elem => elem.display_title.includes(headerRow.display_title))
            .includes(true) === false
        ) {
          idxColumn.push({
            display_title: "",
            isVisible: false,
            key: idxHeader
          });
        }
      });
    });
    return idxColumn.map(column => (
      <td
        id={"td" + index + column.display_title}
        key={"key" + column.key + column.display_title}
        className={column.isVisible ? "checked" : ""}
      >
        <label>
          {column.display_title}
          <input
            id={index + "--" + column.display_title}
            name={index + "--" + column.display_title}
            type="checkbox"
            checked={column.isVisible}
            onChange={this.handleSelecion}
          ></input>
        </label>
      </td>
    ));
  };

  renderSelectColumns = () => {
    this.state.list.forEach(tool => {
      tool.columns.forEach(column => {
        if (
          this.selectable.findIndex(
            value => value.display_title === column.display_title
          ) < 0
        ) {
          this.selectable.push(column);
        }
      });
    });
    return this.selectable.map((column, i) => {
      //column.isVisible === all columns true
      //in all tools => in Tool: column with column.title? => isVisible?
      return (
        <th
          id={"td-all-" + column.display_title}
          key={"key" + column.display_title}
          className={column.isVisible ? "checked" : ""}
        >
          <label>
            {column.display_title}
            <input
              name={column.display_title}
              type="checkbox"
              checked={column.isVisible}
              onChange={this.handleSelectColumns}
            ></input>
          </label>
        </th>
      );
    });
  };

  // -------------------------Handling-------------------------
  handleSelecion = ({ target }) => {
    const [tool, column] = target.name.split("--");
    const value = target.type === "checkbox" ? target.checked : target.value;
    const index = this.state.list[tool].columns.findIndex(
      el => el.display_title === column
    );
    const list = [...this.state.list];

    list[tool].columns[index].isVisible = value;

    this.checkTools(list);
  };

  handleSelectColumns = ({ target }) => {
    const value = target.type === "checkbox" ? target.checked : target.value;
    const list = [...this.state.list];

    list.forEach(tool => {
      const index = tool.columns.findIndex(
        el => el.display_title === target.name
      );
      if (tool.columns[index]) {
        tool.columns[index].isVisible = value;
      }
    });

    this.checkTools(list);
  };
  deselectTool = (i, { target }) => {
    const value = target.type === "checkbox" ? target.checked : target.value;
    const list = [...this.state.list];

    list[i].columns.forEach(column => {
      column.isVisible = value;
    });

    this.checkTools(list);
  };

  deselectAll = () => {
    const list = [...this.state.list];

    list.forEach(tool =>
      tool.columns.forEach(column => {
        column.isVisible = !this.state.deselect;
      })
    );

    this.checkTools(list);
    this.setState(prevState => ({
      deselect: !prevState.deselect
    }));
  };

  checkTools = list => {
    list.forEach(tool => {
      tool.isVisible = tool.columns.findIndex(column => column.isVisible) >= 0;
    });

    this.setState({ list });
  };

  render() {
    ReactModal.setAppElement("#root");
    return (
      <ReactModal
        className="overlay"
        isOpen={true}
        onRequestClose={this.props.close}
      >
        <FontAwesomeIcon
          icon={faTimes}
          onClick={this.props.close}
          className="closing"
        />
        <h1>Select the columns to display</h1>
        <table className="selectRows">
          <tbody>
            <tr className="selectColumn_all">
              <th></th>
              {this.renderSelectColumns()}
            </tr>
            {this.renderRunSets()}
          </tbody>
        </table>
        <div className="overlay__buttons">
          <button className="btn" onClick={this.deselectAll}>
            {this.state.deselect ? "Deselect all" : "Select all"}
          </button>
          <button
            className="btn btn-apply"
            onClick={this.props.close}
            disabled={!this.state.list.filter(tool => tool.isVisible).length}
          >
            Apply and close
          </button>
          <input />
        </div>
      </ReactModal>
    );
  }
}
