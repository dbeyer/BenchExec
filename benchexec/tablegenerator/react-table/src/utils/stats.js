// This file is part of BenchExec, a framework for reliable benchmarking:
// https://github.com/sosy-lab/benchexec
//
// SPDX-FileCopyrightText: 2019-2020 Dirk Beyer <https://www.sosy-lab.org>
//
// SPDX-License-Identifier: Apache-2.0

import { isNil, NumberFormatterBuilder } from "./utils";
import { enqueue } from "../workers/workerDirector";

const keysToIgnore = ["meta"];

export const statisticsRows = {
  total: { title: "all results" },
  correct: {
    indent: 1,
    title: "correct results",
    description:
      "(property holds + result is true) OR (property does not hold + result is false)",
  },
  correct_true: {
    indent: 2,
    title: "correct true",
    description: "property holds + result is true",
  },
  correct_false: {
    indent: 2,
    title: "correct false",
    description: "property does not hold + result is false",
  },
  correct_unconfirmed: {
    indent: 1,
    title: "correct-unconfirmed results",
    description:
      "(property holds + result is true) OR (property does not hold + result is false), but unconfirmed",
  },
  correct_unconfirmed_true: {
    indent: 2,
    title: "correct-unconfirmed true",
    description: "property holds + result is true, but unconfirmed",
  },
  correct_unconfirmed_false: {
    indent: 2,
    title: "correct-unconfirmed false",
    description: "property does not hold + result is false, but unconfirmed",
  },
  wrong: {
    indent: 1,
    title: "incorrect results",
    description:
      "(property holds + result is false) OR (property does not hold + result is true)",
  },
  wrong_true: {
    indent: 2,
    title: "incorrect true",
    description: "property does not hold + result is true",
  },
  wrong_false: {
    indent: 2,
    title: "incorrect false",
    description: "property holds + result is false",
  },
};

/**
 * Remove all statistics rows for which the statistics worker cannot/will not
 * compute values (e.g., local summary, score).
 */
export const filterComputableStatistics = (stats) =>
  stats.filter((row) => statisticsRows[row.id]);

/**
 * This method gets called on the initial render or whenever there is a
 * change to the underlying dataset.
 * This usually happens whenever the user sets a filter.
 *
 * It handles the dispatching of stat calculation jobs as well as
 * necessary transformation to bring the calculation results into the
 * required format.
 */
export const computeStats = async ({ tools, tableData, stats }) => {
  const formatter = buildFormatter(tools);
  let res = await processData({ tools, tableData, formatter });

  const availableStats = stats
    .map((row) => row.id)
    .filter((id) => statisticsRows[id]);
  const cleaned = cleanupStats(res, formatter, availableStats);

  // fill up stat array to match column mapping

  // The result of our stat calculation only contains relevant columns.
  // The stat table however requires a strict ordering of columns that also
  // includes columns that are not even rendered.
  //
  // In order to ensure a consistent layout we iterate through all columns
  // of the runset and append dummy objects until we reach a column that we
  // have calculated data for
  res = cleaned.map((tool, toolIdx) => {
    const out = [];
    const toolColumns = tools[toolIdx].columns;
    let pointer = 0;
    let curr = toolColumns[pointer];

    for (const col of tool) {
      const { title } = col;
      while (pointer < toolColumns.length && title !== curr.title) {
        // irrelevant column
        out.push({});
        pointer++;
        curr = toolColumns[pointer];
      }
      if (pointer >= toolColumns.length) {
        break;
      }
      // relevant column
      out.push(col);
      pointer++;
      curr = toolColumns[pointer];
    }

    return out;
  });

  // Put new statistics in same "shape" as old ones.
  return filterComputableStatistics(stats).map((row) => {
    const content = row.content.map((tool, toolIdx) => {
      return res[toolIdx].map((col) => col[row.id]);
    });
    return { ...row, content };
  });
};

/**
 * Creates a number formatters for each tool and column and
 * configures them with the columns defined number of significant digits
 *
 * @param {object[]} tools
 */
const buildFormatter = (tools) =>
  tools.map((tool, tIdx) =>
    tool.columns.map((column, cIdx) => {
      const { number_of_significant_digits: sigDigits } = column;
      return new NumberFormatterBuilder(sigDigits, `${tIdx}-${cIdx}`);
    }),
  );

const maybeRound =
  (key, maxDecimalInputLength, columnIdx) =>
  (number, { significantDigits }) => {
    const asNumber = Number(number);
    const [integer, decimal] = number.split(".");

    if (["sum", "avg", "stdev"].includes(key)) {
      // for cases when we have no significant digits defined,
      // we want to pad avg and stdev to two digits
      if (isNil(significantDigits) && key !== "sum") {
        return asNumber.toFixed(2);
      }
      // integer value without leading 0
      const cleanedInt = integer.replace(/^0+/, "");
      // decimal value without leading 0, if cleanedInt is empty (evaluates to zero)
      let cleanedDec = decimal || "";
      if (cleanedInt === "") {
        cleanedDec = cleanedDec.replace(/^0+/, "");
      }

      // differences in length between input value with maximal length and current value
      const deltaInputLength = maxDecimalInputLength - (decimal?.length ?? 0);

      // differences in length between num of significant digits and current value
      const deltaSigDigLength =
        significantDigits - (cleanedInt.length + cleanedDec.length);

      // if we have not yet filled the number of significant digits, we could decide to pad
      const paddingPossible = deltaSigDigLength > 0;

      const missingDigits = (decimal?.length ?? 0) + deltaSigDigLength;

      if (deltaInputLength > 0 && paddingPossible && key !== "stdev") {
        if (deltaInputLength > deltaSigDigLength) {
          // we want to pad to the smaller value (sigDigits vs maxDecimal)
          return asNumber.toFixed(missingDigits);
        }
        return asNumber.toFixed(maxDecimalInputLength);
      }

      // if avg was previously padded to fill the number of significant digits,
      // we want to make sure, that we don't go over the maximumDecimalDigits
      if (
        key === "avg" &&
        !paddingPossible &&
        deltaInputLength < 0 &&
        number[number.length - 1] === "0"
      ) {
        return asNumber.toFixed(maxDecimalInputLength);
      }

      if (key === "stdev" && paddingPossible) {
        return asNumber.toFixed(missingDigits);
      }
    }

    return number;
  };

/**
 * Used to apply formatting to calculated stats and to remove
 * values that are not displayable
 *stats
 * @param {object[][]} stats
 * @param {Function[][]} formatter
 */
const cleanupStats = (unfilteredStats, formatter, availableStats) => {
  const stats = unfilteredStats.map((tool, toolIdx) =>
    tool.map((col, colIdx) => {
      const { columnType } = col;
      const out = { columnType };

      for (const visibleStats of availableStats) {
        const currentCol = col[visibleStats];
        if (!currentCol) {
          continue;
        }
        out[visibleStats] = currentCol;
        if (currentCol?.sum ?? false) {
          formatter[toolIdx][colIdx].addDataItem(currentCol.sum);
        }
      }
      return out;
    }),
  );

  for (const to in formatter) {
    for (const co in formatter[to]) {
      // we build all formatters which makes them ready to use
      formatter[to][co] = formatter[to][co].build();
    }
  }

  const cleaned = stats.map((tool, toolIdx) =>
    tool
      .map(({ columnType, ...column }, columnIdx) => {
        const out = {};
        // if no total is calculated, then no values suitable for calculation were found
        if (column.total === undefined) {
          return undefined;
        }
        for (const [resultKey, result] of Object.entries(column)) {
          const rowRes = {};
          const meta = result?.meta;
          for (let [key, value] of Object.entries(result)) {
            // we ignore any of these defined keys
            if (keysToIgnore.includes(key)) {
              continue;
            }

            const maxDecimalInputLength = meta?.maxDecimals ?? 0;

            // attach the title to the stat item
            // this will later be used to ensure correct ordering of columns
            if (key === "title") {
              out.title = value;
              continue;
            }
            // if we have numeric values or 'NaN' we want to apply formatting
            if (
              !isNil(value) &&
              (!isNaN(value) || value === "NaN") &&
              formatter[toolIdx][columnIdx]
            ) {
              try {
                if (key === "sum") {
                  rowRes[key] = formatter[toolIdx][columnIdx](value, {
                    leadingZero: false,
                    whitespaceFormat: true,
                    html: true,
                    additionalFormatting: maybeRound(
                      key,
                      maxDecimalInputLength,
                      columnIdx,
                    ),
                  });
                } else {
                  rowRes[key] = formatter[toolIdx][columnIdx](value, {
                    leadingZero: true,
                    whitespaceFormat: false,
                    html: false,
                    additionalFormatting: maybeRound(
                      key,
                      maxDecimalInputLength,
                      columnIdx,
                    ),
                  });
                }
              } catch (e) {
                console.error({
                  key,
                  value,
                  formatter: formatter[toolIdx][columnIdx],
                  e,
                });
              }
            }
          }

          out[resultKey] = rowRes;
        }

        return out;
      })
      .filter((i) => !isNil(i)),
  );
  return cleaned;
};

const RESULT_TRUE_PROP = "true";
//property holds
const RESULT_FALSE_PROP = "false";
//property does not hold

const RESULT_CLASS_TRUE = "true";
const RESULT_CLASS_FALSE = "false";
const RESULT_CLASS_OTHER = "other";

/**
 * @see result.py
 */
const classifyResult = (result) => {
  if (isNil(result)) {
    return RESULT_CLASS_OTHER;
  }
  if (result === RESULT_TRUE_PROP) {
    return RESULT_CLASS_TRUE;
  }
  if (result === RESULT_FALSE_PROP) {
    return RESULT_CLASS_FALSE;
  }
  if (result.startsWith(`${RESULT_FALSE_PROP}(`) && result.endsWith(")")) {
    return RESULT_CLASS_FALSE;
  }

  return RESULT_CLASS_OTHER;
};

const prepareRows = (
  rows,
  toolIdx,
  categoryAccessor,
  statusAccessor,
  formatter,
) => {
  return rows.map((row) => {
    const cat = categoryAccessor(toolIdx, row);
    const stat = statusAccessor(toolIdx, row);

    const mappedCat = cat.replace(/-/g, "_");
    const mappedRes = classifyResult(stat);

    return {
      categoryType: mappedCat,
      resultType: mappedRes,
      row: row.results[toolIdx].values,
    };
  });
};

/**
 * Transforms the dataset from a row-based layout into a column-based
 * layout for calculation of column stats.
 *
 * @param {object[]} tools
 */
const splitColumnsWithMeta = (tools) => (preppedRows, toolIdx) => {
  const out = [];
  for (const { row, categoryType, resultType } of preppedRows) {
    for (const columnIdx in row) {
      const column = row[columnIdx].raw;
      const curr = out[columnIdx] || [];
      // we attach extra meta information for later use in calculation and mapping
      // of results
      const { type: columnType, title: columnTitle } =
        tools[toolIdx].columns[columnIdx];

      curr.push({ categoryType, resultType, column, columnType, columnTitle });
      out[columnIdx] = curr;
    }
  }
  return out;
};

/**
 * Prepares the dataset for calculation, dispatches and collects calculations
 * and returns a cleaned set of calculated statistics.
 *
 * @param {object} options
 */
const processData = async ({ tools, tableData, formatter }) => {
  const catAccessor = (toolIdx, row) => row.results[toolIdx].category;
  const statAccessor = (toolIdx, row) => row.results[toolIdx].values[0].raw;
  const promises = [];

  const splitRows = [];
  for (const toolIdx in tools) {
    splitRows.push(
      prepareRows(tableData, toolIdx, catAccessor, statAccessor, formatter),
    );
  }
  const columnSplitter = splitColumnsWithMeta(tools);

  const preparedData = splitRows.map(columnSplitter);
  // filter out non-relevant rows
  for (const toolIdx in preparedData) {
    preparedData[toolIdx] = preparedData[toolIdx].filter((i) => !isNil(i));
  }

  for (const toolDataIdx in preparedData) {
    const toolData = preparedData[toolDataIdx];
    const subPromises = [];
    for (const columnIdx in toolData) {
      const columns = toolData[columnIdx];
      subPromises.push(enqueue({ name: "stats", data: columns }));
    }
    promises[toolDataIdx] = subPromises;
  }

  const allPromises = promises.map((p) => Promise.all(p));
  const res = await Promise.all(allPromises);

  return res;
};
