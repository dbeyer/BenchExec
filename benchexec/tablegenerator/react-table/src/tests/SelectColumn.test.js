/* SPDX-License-Identifier: Apache-2.0
 *
 * BenchExec is a framework for reliable benchmarking.
 * This file is part of BenchExec.
 * Copyright (C) Dirk Beyer. All rights reserved.
 */
import React from "react";
import SelectColumn from "../App.js";
import { test_snapshot_of } from "./utils.js";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

Enzyme.configure({ adapter: new Adapter() });

// mock uniqid to have consistent names
// https://stackoverflow.com/a/44538270/396730
jest.mock("uniqid", () => i => i + "uniqid");

const wrapper = shallow(<SelectColumn />);
const rootElement = wrapper.find("#root");
test_snapshot_of("Render App", overview => rootElement);
