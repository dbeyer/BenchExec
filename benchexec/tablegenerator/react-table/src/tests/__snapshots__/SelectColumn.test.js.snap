// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render App for benchmark.test.2019-12-20_1532.results.html 1`] = `null`;

exports[`Render App for benchmark-example-true.2019-11-06_0932.results.no options.html 1`] = `null`;

exports[`Render App for big-table.diff.html 1`] = `null`;

exports[`Render App for cbmc.2015-12-11_1211.results.Simple.html 1`] = `null`;

exports[`Render App for multi-table.diff.html 1`] = `null`;

exports[`Render App for multi-table.table.html 1`] = `null`;

exports[`Render App for multi-table-only-columns.diff.html 1`] = `null`;

exports[`Render App for multi-table-only-columns.table.html 1`] = `null`;

exports[`Render App for multi-table-with-columns.diff.html 1`] = `null`;

exports[`Render App for multi-table-with-columns.table.html 1`] = `null`;

exports[`Render App for multi-table-with-diff-over-column.diff.html 1`] = `null`;

exports[`Render App for multi-table-with-diff-over-column.table.html 1`] = `null`;

exports[`Render App for multi-table-with-wildcards.diff.html 1`] = `null`;

exports[`Render App for multi-table-with-wildcards.table.html 1`] = `null`;

exports[`Render App for nan_and_inf.html 1`] = `null`;

exports[`Render App for predicateAnalysis.table.html 1`] = `null`;

exports[`Render App for predicateAnalysis-reverse.table.html 1`] = `null`;

exports[`Render App for simple-table-with-columns.table.html 1`] = `null`;

exports[`Render App for simple-table-with-links.table.html 1`] = `null`;

exports[`Render App for simple-table-with-numberOfDigits.table.html 1`] = `null`;

exports[`Render App for simple-table-with-scaling.table.html 1`] = `null`;

exports[`Render App for smt.diff.html 1`] = `null`;

exports[`Render App for smt.table.html 1`] = `null`;

exports[`Render App for test.2015-03-03_1613.diff.html 1`] = `null`;

exports[`Render App for test.2015-03-03_1613.results.predicateAnalysis.all-columns.html 1`] = `null`;

exports[`Render App for test.2015-03-03_1613.results.predicateAnalysis.correct-only.html 1`] = `null`;

exports[`Render App for test.2015-03-03_1613.results.predicateAnalysis.custom-score.html 1`] = `null`;

exports[`Render App for test.2015-03-03_1613.results.predicateAnalysis.html 1`] = `null`;

exports[`Render App for test.2015-03-03_1613.results.valueAnalysis.html 1`] = `null`;

exports[`Render App for test.2015-03-03_1613.results.valueAnalysis.table.html 1`] = `null`;

exports[`Render App for test.2015-03-03_1613.table.html 1`] = `null`;

exports[`Render App for test.2015-03-03_1613-common.diff.html 1`] = `null`;

exports[`Render App for test.2015-03-03_1613-common.table.html 1`] = `null`;

exports[`Render App for test.2015-03-03_1613-correct-only.diff.html 1`] = `null`;

exports[`Render App for test.2015-03-03_1613-correct-only.table.html 1`] = `null`;

exports[`Render App for test.2015-03-03_1613-reverse.diff.html 1`] = `null`;

exports[`Render App for test.2015-03-03_1613-reverse.table.html 1`] = `null`;

exports[`Render App for test-error.2015-03-03_1613.diff.html 1`] = `null`;

exports[`Render App for test-error.2015-03-03_1613.table.html 1`] = `null`;

exports[`Render App for union-table.diff.html 1`] = `null`;

exports[`Render App for union-table.table.html 1`] = `null`;

exports[`Render App for union-table-duplicate-results.diff.html 1`] = `null`;

exports[`Render App for union-table-duplicate-results.table.html 1`] = `null`;

exports[`Render App for union-table-mixed.diff.html 1`] = `null`;

exports[`Render App for union-table-mixed.table.html 1`] = `null`;

exports[`Render App for union-table-multiple-results.diff.html 1`] = `null`;

exports[`Render App for union-table-multiple-results.table.html 1`] = `null`;

exports[`Render App for union-table-predicateAnalysis.table.html 1`] = `null`;

exports[`Render App for union-table-valueAnalysis.table.html 1`] = `null`;
